crates/bevy_render/src/pipeline_keys/mod.rs

```diff
@@ -0,0 +1,682 @@
+use crate::{render_resource::ShaderDefVal, *};
+use bevy_app::prelude::*;
+use bevy_ecs::{prelude::*, query::*, schedule::NodeConfigs, system::*};
+pub use bevy_render_macros::PipelineKey;
+use bevy_utils::{intern::Interned, HashMap, HashSet};
+use std::{
+    any::{type_name, TypeId},
+    marker::PhantomData,
+};
+
+mod composite;
+mod packed_types;
+
+/// A pipeline key is a struct or enum implementing the `PipelineKeyType` trait, which provides methods to convert it
+/// into a packed set of bits in a `PackedPipelineKey` struct, and back again.
+/// Pipeline keys are used to cache pipeline variants, so it must be possible to obtain all information that can change
+/// the pipeline (shader defs, bindings, outputs, entry points, etc) from the pipeline's key value.
+/// Keys may specify a set of `ShaderDefVal`s (based on the key value) that can be applied to the vertex and fragment states
+/// of a `RenderPipelineDescriptor` or `ComputePipelineDescriptor`, via the `KeyShaderDefs` trait.
+/// Key values can be calculated independently for a given entity via the `SystemKey` trait.
+/// Keys can also be composites of other keys, either as simple tuples, as structs, or as dynamic keys, the components of which
+/// can be independently registered to allow extensibility.
+
+/// The primitive type underlying packed pipeline keys.
+pub type KeyPrimitive = u128;
```

Comment:
So each key will have 128 bits? I recall there were concerns regarding hashing perf.

Reply:
It could be 64 but with some compromises on current data - specifically the StandardMaterial data uses an i32 for depth bias which could comfortably be condensed. The view and mesh keys are around 30 bits all in

Reply:
i don't know all the details, but i'm sure there'll be some hardware where it's as fast and some where it's not.